File: tanuki_checker/passes/lexical.py
import re
import math
from typing import Dict, List
from tanuki_checker.ast import DocumentAST, NodeType, TokenTag
from tanuki_checker.symbol import SymbolTable
from tanuki_checker.passes.base import IAnalysisPass, PassExecutionError
from tanuki_checker.metrics import LexicalMetrics

AI_FREQUENT_PHRASES = [
    r"徹底解説",
    r"以下のポイントに留意",
    r"まとめとして",
    r"をはじめとする",
    r"多角的な視点から",
    r"重要な役割を果たし",
    r"詳しく見ていきましょう",
    r"背景には.*?挙げられます",
    r"注目を集めています",
    r"本記事では",
    r"一概には言えませんが",
    r"不可欠な要素",
    r"実践的な",
    r"注力することが推奨されます",
    r"環境を整備することが重要",
    r"ご参考になりましたら幸いです",
    r"お気軽にお知らせください",
    r"以下のメリット",
    r"を享受",
    r"是非.*?ご検討ください",
    r"結論から言うと",
    r"結局のところ",
    r"これからの時代",
    r"大切になってくる",
    r"コメントで教えて",
    r"\bin conclusion\b",
    r"\bit is worth noting\b",
    r"\bdelve into\b",
    r"\bplays a crucial role\b",
    r"\btestament to\b",
    r"\btapestry of\b",
    r"\bin today's fast-paced world\b",
]

HEDGE_EXPRESSIONS = [
    r"と思われる",
    r"と考えられる",
    r"一説には",
    r"一般的には",
    r"と言えるでしょう",
    r"推測される",
    r"見込まれる",
    r"可能性が高い",
    r"\bit is generally believed\b",
    r"\bit seems that\b",
    r"\barguably\b",
    r"\bmay be considered\b",
    r"\bcould be argued\b",
]


class LexicalPass(IAnalysisPass):
    def __init__(self):
        self._ai_phrase_res = [re.compile(p, re.IGNORECASE) for p in AI_FREQUENT_PHRASES]
        self._hedge_res = [re.compile(p, re.IGNORECASE) for p in HEDGE_EXPRESSIONS]

    @property
    def name(self) -> str:
        return "LexicalPass"

    def execute(self, ast: DocumentAST, sym_table: SymbolTable) -> Dict[str, float]:
        try:
            metrics = self.analyze(ast, sym_table)
            return metrics.model_dump()
        except Exception as e:
            raise PassExecutionError(f"Error in LexicalPass: {str(e)}") from e

    def analyze(self, ast: DocumentAST, sym_table: SymbolTable) -> LexicalMetrics:
        doc_node = ast.get_node(ast.root_id)
        if not doc_node or not doc_node.raw_text:
            return LexicalMetrics()

        text = doc_node.raw_text
        text_len = max(len(text), 1)

        # 1. AI Phrase Density
        ai_phrase_count = sum(len(r.findall(text)) for r in self._ai_phrase_res)

        # Density per 1000 characters
        ai_phrase_density = (ai_phrase_count / text_len) * 1000.0

        # 2. Hedge Expression Ratio
        hedge_count = sum(len(r.findall(text)) for r in self._hedge_res)

        sentence_nodes = ast.get_nodes_by_type(NodeType.SENTENCE)
        total_sentences = max(len(sentence_nodes), 1)
        hedge_expression_ratio = min(hedge_count / total_sentences, 1.0)

        # 3. N-gram Entropy (Character 2-gram entropy)
        ngram_entropy = self._calculate_ngram_entropy(text, n=2)

        return LexicalMetrics(
            ai_phrase_density=round(ai_phrase_density, 4),
            ngram_entropy=round(ngram_entropy, 4),
            hedge_expression_ratio=round(hedge_expression_ratio, 4),
        )

    def _calculate_ngram_entropy(self, text: str, n: int = 2) -> float:
        cleaned_text = re.sub(r"\s+", "", text)
        if len(cleaned_text) < n:
            return 0.0

        ngrams: Dict[str, int] = {}
        total_ngrams = len(cleaned_text) - n + 1
        for i in range(total_ngrams):
            gram = cleaned_text[i : i + n]
            ngrams[gram] = ngrams.get(gram, 0) + 1

        entropy = 0.0
        for count in ngrams.values():
            p = count / total_ngrams
            entropy -= p * math.log2(p)

        return entropy

File: src/analyzer/passes/lexical.js
import { NodeType } from '../ast.js';

const AI_FREQUENT_PHRASES = [
  "徹底解説",
  "以下のポイントに留意",
  "まとめとして",
  "をはじめとする",
  "多角的な視点から",
  "重要な役割を果たし",
  "詳しく見ていきましょう",
  "背景には.*?挙げられます",
  "注目を集めています",
  "本記事では",
  "一概には言えませんが",
  "不可欠な要素",
  "実践的な",
  "注力することが推奨されます",
  "環境を整備することが重要",
  "ご参考になりましたら幸いです",
  "お気軽にお知らせください",
  "以下のメリット",
  "を享受",
  "是非.*?ご検討ください",
  "結論から言うと",
  "結局のところ",
  "これからの時代",
  "大切になってくる",
  "コメントで教えて",
  "\\bin conclusion\\b",
  "\\bit is worth noting\\b",
  "\\bdelve into\\b",
  "\\bplays a crucial role\\b",
  "\\btestament to\\b",
  "\\btapestry of\\b",
  "\\bin today's fast-paced world\\b",
];

const HEDGE_EXPRESSIONS = [
  "と思われる",
  "と考えられる",
  "一説には",
  "一般的には",
  "と言えるでしょう",
  "推測される",
  "見込まれる",
  "可能性が高い",
  "\\bit is generally believed\\b",
  "\\bit seems that\\b",
  "\\barguably\\b",
  "\\bmay be considered\\b",
  "\\bcould be argued\\b",
];

export class LexicalPass {
  constructor() {
    this._aiPhraseRes = AI_FREQUENT_PHRASES.map(p => new RegExp(p, 'gi'));
    this._hedgeRes = HEDGE_EXPRESSIONS.map(p => new RegExp(p, 'gi'));
  }

  get name() {
    return 'LexicalPass';
  }

  execute(ast, symTable) {
    const docNode = ast.get_node(ast.root_id);
    if (!docNode || !docNode.raw_text) {
      return {
        ai_phrase_density: 0.0,
        ngram_entropy: 0.0,
        hedge_expression_ratio: 0.0
      };
    }

    const text = docNode.raw_text;
    const textLen = Math.max(text.length, 1);

    // 1. AI Phrase Density
    let aiPhraseCount = 0;
    for (const r of this._aiPhraseRes) {
      const matches = text.match(r);
      if (matches) aiPhraseCount += matches.length;
    }
    const aiPhraseDensity = (aiPhraseCount / textLen) * 1000.0;

    // 2. Hedge Expression Ratio
    let hedgeCount = 0;
    for (const r of this._hedgeRes) {
      const matches = text.match(r);
      if (matches) hedgeCount += matches.length;
    }
    const sentenceNodes = ast.get_nodes_by_type(NodeType.SENTENCE);
    const totalSentences = Math.max(sentenceNodes.length, 1);
    const hedgeExpressionRatio = Math.min(hedgeCount / totalSentences, 1.0);

    // 3. N-gram Entropy (Character 2-gram entropy)
    const ngramEntropy = this._calculateNgramEntropy(text, 2);

    return {
      ai_phrase_density: Number(aiPhraseDensity.toFixed(4)),
      ngram_entropy: Number(ngramEntropy.toFixed(4)),
      hedge_expression_ratio: Number(hedgeExpressionRatio.toFixed(4))
    };
  }

  _calculateNgramEntropy(text, n = 2) {
    const cleanedText = text.replace(/\s+/g, '');
    if (cleanedText.length < n) return 0.0;

    const ngrams = {};
    const totalNgrams = cleanedText.length - n + 1;
    for (let i = 0; i < totalNgrams; i++) {
      const gram = cleanedText.slice(i, i + n);
      ngrams[gram] = (ngrams[gram] || 0) + 1;
    }

    let entropy = 0.0;
    for (const count of Object.values(ngrams)) {
      const p = count / totalNgrams;
      entropy -= p * Math.log2(p);
    }

    return entropy;
  }
}
